packages/core/src/api/Editor.ts

```diff
@@ -0,0 +1,433 @@
+          "blockGroup"
+        ].create({}, nodesToInsert);
+
+        this.tiptapEditor.view.dispatch(
+          this.tiptapEditor.state.tr.insert(insertionPos, blockGroupNode)
+        );
+
+        return;
+      }
+
+      insertionPos = posBeforeNode + node.firstChild!.nodeSize + 2;
+    }
+
+    this.tiptapEditor.view.dispatch(
+      this.tiptapEditor.state.tr.insert(insertionPos, nodesToInsert)
+    );
+  }
+
+  /**
+   * Updates a block in the editor to the given specification.
+   * @param blockToUpdate The block that should be updated.
+   * @param updatedBlock The specification that the block should be updated to.
+   */
+  public updateBlock(blockToUpdate: Block, updatedBlock: PartialBlock) {
+    const { posBeforeNode } = this.getNodeById(blockToUpdate.id);
+
+    this.tiptapEditor.commands.BNUpdateBlock(posBeforeNode + 1, updatedBlock);
+  }
+
+  /**
+   * Removes multiple blocks from the editor. Throws an error if any of the blocks could not be found.
+   * @param blocksToRemove An array of blocks that should be removed.
+   */
+  public removeBlocks(blocksToRemove: Block[]) {
+    const idsOfBlocksToRemove = new Set<string>(
+      blocksToRemove.map((block) => block.id)
+    );
+
+    let removedSize = 0;
+
+    this.tiptapEditor.state.doc.descendants((node, pos) => {
+      // Skips traversing nodes after all target blocks have been removed.
+      if (idsOfBlocksToRemove.size === 0) {
+        return false;
+      }
+
+      // Keeps traversing nodes if block with target ID has not been found.
+      if (
+        node.type.name !== "blockContainer" ||
+        !idsOfBlocksToRemove.has(node.attrs.id)
+      ) {
+        return true;
+      }
+
+      idsOfBlocksToRemove.delete(node.attrs.id);
+      const oldDocSize = this.tiptapEditor.state.doc.nodeSize;
+
+      this.tiptapEditor.commands.BNDeleteBlock(pos - removedSize + 1);
+
+      const newDocSize = this.tiptapEditor.state.doc.nodeSize;
+      removedSize += oldDocSize - newDocSize;
+
+      return false;
+    });
+
+    if (idsOfBlocksToRemove.size > 0) {
+      let notFoundIds = "";
+
+      for (const id of Array.from(idsOfBlocksToRemove)) {
+        notFoundIds += "\n" + id;
+      }
+
+      throw Error(
+        "Blocks with the following IDs could not be found in the editor: " +
+          notFoundIds
+      );
+    }
+  }
+
+  /**
+   * Replaces multiple blocks in the editor with several other blocks. If the provided blocks to remove are not adjacent
+   * to each other, the new blocks are inserted at the position of the first block in the array. Throws an error if any
+   * of the blocks could not be found.
+   * @param blocksToRemove An array of blocks that should be replaced.
+   * @param blocksToInsert An array of blocks to replace the old ones with.
+   */
+  public replaceBlocks(
+    blocksToRemove: Block[],
+    blocksToInsert: PartialBlock[]
+  ) {
+    this.insertBlocks(blocksToInsert, blocksToRemove[0]);
```

Comment:
Let's say we call this API. What happens if we hit Cmd+Z for undo after this?

Reply:
Seems to work fine, reverts the both inserting the new blocks and removing the old ones.

Reply:
Interesting, that surprises me. How does PM know it's "1 transaction"?

Reply:
notes after call: for now it works (PM-history automatically groups transactions), but might not work with yjs plugin or in other specific scenarios